### PR TITLE
Loose coupling of RiotApi

### PR DESCRIPTION
**Pull request by Oucema90**
#468

### CODE REVIEW

**Comment by BenFradet** [comment]
I'm about half way through the review, this is awesome so far 👍

File: RiotSharp/Endpoints/ChampionMasteryEndpoint/ChampionMasteryEndpointImp.cs [hunk 1]
```diff
@@ -0,0 +1,72 @@
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Newtonsoft.Json;
+using RiotSharp.Endpoints.Interfaces;
+using RiotSharp.Http.Interfaces;
+using RiotSharp.Misc;
+
+namespace RiotSharp.Endpoints.ChampionMasteryEndpoint
+{
+    class ChampionMasteryEndpointImp : IChampionMasteryEndpoint
```
Comment (JanOuborny) [inline RiotSharp/Endpoints/ChampionMasteryEndpoint/ChampionMasteryEndpointImp.cs]: Could you rename this to `ChampionMasteryEndpoint` without `Imp`.

File: RiotSharp/Endpoints/LeagueEndpoint/LeagueEndpointImp.cs [hunk 2]
```diff
@@ -0,0 +1,67 @@
+﻿using System.Collections.Generic;
+using System.Threading.Tasks;
+using Newtonsoft.Json;
+using RiotSharp.Endpoints.Interfaces;
+using RiotSharp.Http.Interfaces;
+using RiotSharp.Misc;
+
+namespace RiotSharp.Endpoints.LeagueEndpoint
+{
+    public class LeagueEndpointImp : ILeagueEndpoint
```
Comment (JanOuborny) [inline RiotSharp/Endpoints/LeagueEndpoint/LeagueEndpointImp.cs]: same here

File: RiotSharp/Endpoints/MasteriesEndpointImp.cs [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+using System.Threading.Tasks;
+using Newtonsoft.Json;
+using RiotSharp.Endpoints.Interfaces;
+using RiotSharp.Endpoints.SummonerEndpoint;
+using RiotSharp.Http.Interfaces;
+using RiotSharp.Misc;
+
+namespace RiotSharp.Endpoints
+{
+    public class MasteriesEndpointImp : IMasteriesEndpoint
```
Comment (JanOuborny) [inline RiotSharp/Endpoints/MasteriesEndpointImp.cs]: same here

File: RiotSharp/Endpoints/MatchEndpoint/MatchEndpointImp.cs [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+using System.Threading.Tasks;
+using Newtonsoft.Json;
+using RiotSharp.Endpoints.Interfaces;
+using RiotSharp.Endpoints.MatchEndpoint.Enums;
+using RiotSharp.Http.Interfaces;
+using RiotSharp.Misc;
+
+namespace RiotSharp.Endpoints.MatchEndpoint
+{
+    class MatchEndpointImp : IMatchEndpoint
```
Comment (JanOuborny) [inline RiotSharp/Endpoints/MatchEndpoint/MatchEndpointImp.cs]: same here

File: RiotSharp/Endpoints/RunesEndpoint/RunesEndpointImp.cs [hunk 5]
```diff
@@ -0,0 +1,38 @@
+﻿using System.Collections.Generic;
+using System.Threading.Tasks;
+using Newtonsoft.Json;
+using RiotSharp.Endpoints.Interfaces;
+using RiotSharp.Http.Interfaces;
+using RiotSharp.Misc;
+
+namespace RiotSharp.Endpoints.RunesEndpoint
+{
+    class RunesEndpointImp : IRunesEndpoint
```
Comment (JanOuborny) [inline RiotSharp/Endpoints/RunesEndpoint/RunesEndpointImp.cs]: same here

File: RiotSharp/Endpoints/SpectatorEndpoint/SpectatorEndpointImp.cs [hunk 6]
```diff
@@ -0,0 +1,47 @@
+﻿using System.Threading.Tasks;
+using Newtonsoft.Json;
+using RiotSharp.Endpoints.Interfaces;
+using RiotSharp.Http.Interfaces;
+using RiotSharp.Misc;
+
+namespace RiotSharp.Endpoints.SpectatorEndpoint
+{
+    public class SpectatorEndpointImp : ISpectatorEndpoint
```
Comment (JanOuborny) [inline RiotSharp/Endpoints/SpectatorEndpoint/SpectatorEndpointImp.cs]: same here

File: RiotSharp/Endpoints/SummonerEndpoint/SummonerEndpointImp.cs [hunk 7]
```diff
@@ -0,0 +1,95 @@
+﻿using System.Threading.Tasks;
+using Newtonsoft.Json;
+using RiotSharp.Endpoints.Interfaces;
+using RiotSharp.Http.Interfaces;
+using RiotSharp.Misc;
+
+namespace RiotSharp.Endpoints.SummonerEndpoint
+{
+    public class SummonerEndpointImp : ISummonerEndpoint
```
Comment (JanOuborny) [inline RiotSharp/Endpoints/SummonerEndpoint/SummonerEndpointImp.cs]: same here

**Comment by Oucema90** [comment]
Thanks for your positive feedbacks. @JanOuborny  @BenFradet 

I have updated the PR to apply the required changes regarding this comments 

> Could you rename this to ChampionMasteryEndpoint without Imp.
> Couldn't you just inherit all methods from the endpoint interfaces?

For this one 

> Could we go even further and just let people access it through IRiotApi.Champion.GetChampions(region) it's available now for all endpoints except for SummonerEndpoint and MatchEndpoint since they are using cache within RiotApi.

**Comment by BenFradet** [comment]
Cool, thanks 👍 .

Last question from me: why did you leverage moq only for the runes and spectator endpoints?

**Comment by Oucema90** [comment]
I leveraged moq only for the runes and spectator endpoints for to test if everything is well with the changes I have made. I'm planning to moq the other unit tests within #459. If you request I finish it within this PR I will start writing the missing ones.

**Comment by BenFradet** [comment]
@Oucema90 no, it's totally fine to separate the changes (it made for an easier review anyway), I was just curious.

It's good to go for me!

Waiting for @JanOuborny to re-review.

**Comment by JanOuborny** [comment]
> it's available now for all endpoints except for SummonerEndpoint and MatchEndpoint since they are using cache within RiotApi.

@Oucema90 
Why can't you inject the `ICache` interface into the `SummonerEndpoint` and `MatchEndpoint` and implement the method in the same way like the other endpoints?

**Comment by Oucema90** [comment]
@JanOuborny PR updated. Thanks for your patience.

File: RiotSharp/RiotApi.cs [hunk 8]
```diff
@@ -122,19 +86,27 @@ public static RiotApi GetInstance(string apiKey, int rateLimitPer10s, int rateLi
...
-            return instance;
+            return _instance;
         }
 
         private RiotApi(string apiKey, IDictionary<TimeSpan, int> rateLimits)
         {
             Requesters.RiotApiRequester = new RateLimitedRequester(apiKey, rateLimits);
-            requester = Requesters.RiotApiRequester;
+            var requester = Requesters.RiotApiRequester;
+            Summoner = new SummonerEndpoint(requester,_cache);
```
Comment (JanOuborny) [inline RiotSharp/RiotApi.cs]: Space missing between arguments

File: RiotSharp/RiotApi.cs [hunk 9]
```diff
@@ -122,19 +86,27 @@ public static RiotApi GetInstance(string apiKey, int rateLimitPer10s, int rateLi
...
+            Champion = new ChampionEndpoint(requester);
+            Masteries = new MasteriesEndpoint(requester);
+            Runes = new RunesEndpoint(requester);
+            League = new LeagueEndpoint(requester);
+            Match = new MatchEndpoint(requester,_cache);
```
Comment (JanOuborny) [inline RiotSharp/RiotApi.cs]: same here

File: RiotSharp/RiotApi.cs [hunk 10]
```diff
@@ -144,486 +116,18 @@ private RiotApi(string apiKey, IDictionary<TimeSpan, int> rateLimits)
...
-                }
+                throw new ArgumentNullException(nameof(rateLimitedRequester));
             }
-            return addedArguments;
+           Summoner = new SummonerEndpoint(rateLimitedRequester, _cache);
+           Champion = new ChampionEndpoint(rateLimitedRequester);
+           Masteries = new MasteriesEndpoint(rateLimitedRequester);
+           Runes = new RunesEndpoint(rateLimitedRequester);
+           League = new LeagueEndpoint(rateLimitedRequester);
+           Match = new MatchEndpoint(rateLimitedRequester,_cache);
```
Comment (JanOuborny) [inline RiotSharp/RiotApi.cs]: same here

**Comment by Oucema90** [comment]
@JanOuborny PR again updated.

**Comment by BenFradet** [comment]
Merged, thank you so much!

**Pull request merged**